html/user/openssl_sign_action.php

```diff
@@ -0,0 +1,148 @@
+//            tra("Your reCAPTCHA response was not correct. Please try again.")
+//        );
+//    }
+//}
+
+// Input is passed in from the openssl_sign_form
+// sanitize_tags removes html elements from the user input
+// $user_data = htmlspecialchars((post_str("user_data", true)),ENT_XML1);
+
+// user data input without htmlspecialchars nor sanitize_tags
+$user_data = post_str("user_data", true);
+
+//Has the user entered data with length > 0 & <= 256?
+if ((strlen($user_data) > 0) && (strlen($user_data) <= 256)) {
+
+	// Retrieve the logged in user's UserID.
+	$userid = $user->id;
+	//$cpid = $user->cross_project_id;
+
+	// Create the message which will be signed.
+	$message_data = "$userid $user_data";
+
+	// Could add extra info into the message however this may cause issues for external systems.
+	// $timestamp = time();
+	// $message_data = "$userid $user_data $timestamp $cpid";
+
+	// How to generate required keys in /project/keys/ folder:
+	// openssl genpkey -algorithm RSA -out external_system_key_private.pem -pkeyopt rsa_keygen_bits:2048
+	// openssl rsa -pubout -in external_system_key_private.pem -out external_system_key_public.pem
+
+	// FAO project admins: Change the following 2 lines to the location of the keys you have personally generated.
+	// If the following keys do not exist, then the users will face (handled) error messages.
+	$private_key_path = "file:///root/project/keys/external_system_key_private.pem";
+	$public_key_path = "file:///root/project/keys/external_system_key_public.pem";
+
+	// Check that the private key file exists where specified. If not, redirect to error page.
+	if (!file_exists($private_key_path)) {
+		error_page(tra("The private key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Check that the public key file exists where specified. If not, redirect to error page.
+	if (!file_exists($public_key_path)) {
+		error_page(tra("The public key does not exist. Proof of UserID currently not possible."));
+	}
+
+	// Fetch private key from file
+	$private_key_pem = openssl_pkey_get_private($private_key_path);
+
+	// Compute signature using the SHA512 signing algorithm
+	$openssl_signed_message = openssl_sign($message_data, $signature, $private_key_pem, OPENSSL_ALGO_SHA512);
+
+	// Free the private key from memory
+	openssl_free_key($private_key_pem);
+	
+	// Check that the signature exists
+	if ($openssl_signed_message == 1) {
+		// The signature was successfully created.
+
+		// fetch public key into memory
+		$pubkeyid = openssl_pkey_get_public($public_key_path);
+
+		// Base64 encode the generated signature to enable safe output to text file.
+		$base64_sig = base64_encode($signature);
+
+		// Decode base64 sig for use in sig_verification
+		$decoded_sig = base64_decode($base64_sig);
+
+		// Verify that the generated signature against the original data, using the public key.
+		$sig_verification = openssl_verify($message_data, $decoded_sig, $pubkeyid, OPENSSL_ALGO_SHA512);
```

Comment:
I do not think it is necessary to verify the signature right after making it. It is one more cryptographic operation on server side. This should be moved behind a condition that can be enabled for debugging/testing and disabled for production.

Reply:
The current performance is near instant, so removing the check doesn't deteriorate user experience. I do agree that it's somewhat unnecessary, however we did experience signature errors with the older openssl version. Commenting it out for debugging/testing wouldn't take much effort to be fair.